libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,729 @@
+		default:
+			throwBadInstruction();
+		}
+
+	case 3:
+
+		switch (u)
+		{
+		case 0:
+			for (int i = 0; i < m; ++i)
+				v8x32 (m_SPP[0])[i] = v64x4(m_SP[1])[v8x32 (m_SP[0])[i] % 32];
+			break;
+		case 1:
+			for (int i = 0; i < m; ++i)
+				v16x16(m_SPP[0])[i] = v64x4(m_SP[1])[v16x16(m_SP[0])[i] % 16];
+			break;
+		case 2:
+			for (int i = 0; i < m; ++i)
+				v32x8 (m_SPP[0])[i] = v64x4(m_SP[1])[v32x8 (m_SP[0])[i] %  8];
+			break;
+		case 3:
+			for (int i = 0; i < m; ++i)
+				v64x4 (m_SPP[0])[i] = v64x4(m_SP[1])[v64x4 (m_SP[0])[i] %  4];
+			break;
+		default:
+			throwBadInstruction();
+		}
+
+	default:
+		throwBadInstruction();
+	}
+}
+
+void VM::xput(uint8_t _b, uint8_t _c)
+{
+	// n type t elements in source and destination vectors, m type u elements in put index
+	uint8_t const t = elemType(_b);
+	uint8_t const m = nElem(_c);
+	uint8_t const u = elemType(_c);
+
+	// given the type of the source, destination and index
+	// for every element of the index put the indexed replacement in the destination                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
+	switch (t)
+	{
+	case 0:
+
+		switch (u)
+		{
+		case 0:
+			for (int i = 0; i < m; ++i)
+				v8x32 (m_SPP[0])[v8x32(m_SP[1])[i] % 32] = v8x32(m_SP[0])[i];
+			break;
+		case 1:
+			for (int i = 0; i < m; ++i)
+				v16x16(m_SPP[0])[v8x32(m_SP[1])[i] % 16] = v8x32(m_SP[0])[i];
```

Comment:
Here I'm still confused, it reads only two vectors from the stack, but XPUT is supposed to take 3 vectors?